datafusion/physical-expr/src/functions.rs

```diff
@@ -253,6 +253,28 @@ macro_rules! invoke_if_unicode_expressions_feature_flag {
 /// decorates a function to handle [`ScalarValue`]s by converting them to arrays before calling the function
 /// and vice-versa after evaluation.
 pub fn make_scalar_function<F>(inner: F) -> ScalarFunctionImplementation
+where
+    F: Fn(&[ArrayRef]) -> Result<ArrayRef> + Sync + Send + 'static,
+{
+    make_scalar_function_with_hints(inner, vec![])
+}
+
+/// Just like [`make_scalar_function`], decorates the given function to handle both [`ScalarValue`]s and arrays.
+/// Additionally can receive a `hints` vector which can be used to control the output arrays when generating them
+/// from [`ScalarValue`]s.
+///
+/// Each element of the `hints` vector gets mapped to the corresponding argument of the function. The number of hints
+/// can be less or greater than the number of arguments (for functions with variable number of arguments). Each unmapped
+/// argument will assume the default hint.
+///
+/// Hints:
```

Comment:
I suggest changing this from `true` / `false` to some sort of enum to make the code potentially easier to read (not having to remember what true / false mean)

So for example, perhaps something like

```rust
enum Hint {
  /// Indicates the argument needs to be padded if it is scalar
  Pad,
  /// indicates the argument can be converted to an array of length 1
  ToArray
}
```

Or something like that

Reply:
I was initially going for a rather simple approach, but I think it makes sense to at least use enums, thanks @alamb!